master/buildbot/www/hooks/bitbucketserver.py

```diff
@@ -0,0 +1,164 @@
+# This file is part of Buildbot.  Buildbot is free software: you can
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+# Copyright Mamba Team
+
+from __future__ import absolute_import
+from __future__ import print_function
+
+import json
+
+from twisted.python import log
+
+GIT_BRANCH_REF = "refs/heads/{}"
+GIT_MERGE_REF = "refs/pull-requests/{}/merge"
+GIT_TAG_REF = "refs/tags/{}"
+
+_HEADER_EVENT = 'X-Event-Key'
+
+
+class BitbucketServerEventHandler(object):
```

Comment:
Please have a look at https://github.com/buildbot/buildbot/pull/3327
It is a refactor to make the hooks more maintainable and flexible. you will need to rework that part when it lands